lib/codegen/src/nan_canonicalization.rs

```diff
@@ -0,0 +1,81 @@
+//! A NaN-canonicalizing rewriting pass. Patch floating point arithmetic
+//! instructions that may return a NaN result with a sequence of operations
+//! that will replace nondeterministic NaN's with a single canonical NaN value.
+
+use cursor::{Cursor, FuncCursor};
+use ir::{Function, Inst, InstBuilder, InstructionData, Opcode, Value};
+use ir::condcodes::FloatCC;
+use ir::immediates::{Ieee32, Ieee64};
+use ir::types;
+use ir::types::Type;
+use timing;
+
+// Canonical 32-bit and 64-bit NaN values.
+static CANON_32BIT_NAN: u32 = 0b01111111100000000000000000000001;
+static CANON_64BIT_NAN: u64 = 0b0111111111110000000000000000000000000000000000000000000000000001;
```

Comment:
In case github swallowed my comment here, we need to use quiet NaN bit-patterns here, so the most significant bit of the significand field needs to be 1. And then, we should make the last significant bit 0, because that matches what several other systems consider to be the canonical NaN bit pattern.